### PR TITLE
[ONNX] Convolution operation.

### PR DESCRIPTION
**Pull request by arogowie-intel**
Please merge after #1377 is merged.

### CODE REVIEW

File: src/ngraph/frontend/onnx_import/model.hpp [hunk 1]
```diff
@@ -17,8 +17,8 @@
 #pragma once
 
 #include <ostream>
-#include "onnx.pb.h"
 #include "onnx-operators.pb.h"
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/model.hpp]: Why do we need to include this?

File: src/ngraph/frontend/onnx_import/op/conv.hpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                inline std::shared_ptr<ngraph::op::Op>
+                    make_ng_convolution(const std::shared_ptr<ngraph::Node>& data,
+                                        const std::shared_ptr<ngraph::Node>& filters,
+                                        const ngraph::Strides& strides,
+                                        const ngraph::Strides& dilations,
+                                        const ngraph::CoordinateDiff& padding_below,
+                                        const ngraph::CoordinateDiff& padding_above,
+                                        int groups)
+                {
+                    if (groups != 1)
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/conv.hpp]: Implementation should be moved to .cpp file. Please review nGraph coding conventions.

File: src/ngraph/frontend/onnx_import/CMakeLists.txt [hunk 3]
```diff
@@ -56,6 +56,7 @@ add_library(onnx_import STATIC
     node.hpp
     op/add.hpp
     op/constant.hpp
+    op/conv.hpp
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/CMakeLists.txt]: Should we add exceptions.hpp here?

Comment (arogowie-intel) [reply src/ngraph/frontend/onnx_import/CMakeLists.txt]: Forgot about this. Thanks.

**Comment by arogowie-intel** [comment]
Have some problems with generating correct model for testing convolution with bias. Don't know why the code below generates model with no outputs, even though the graph contains output as you may see in the script console output.
```python
import onnx
from onnx.helper import make_node, make_graph, make_tensor_value_info, make_model
from string import ascii_uppercase
import numpy as np

def get_node_model(op_type, output_shapes, *input_data,  opset=1, 
      num_outputs=1, **node_attributes):
   node_inputs = [np.array(data) for data in input_data]
   num_inputs = len(node_inputs)
   node_input_names = [ascii_uppercase[idx] for idx in range(num_inputs)]
   node_output_names = [ascii_uppercase[num_inputs + idx] for idx in range(num_outputs)]

   print('inputs: ' + str([str(n) for n in node_input_names]))
   print('outputs: ' + str([str(n) for n in node_output_names]))
   onnx_node = make_node(op_type, node_input_names, node_output_names, **node_attributes)
   print('onnx_node: ' + str(onnx_node))
   input_tensors = [make_tensor_value_info(name, onnx.TensorProto.FLOAT, value.shape)
                       for name, value in zip(onnx_node.input, node_inputs)]
   output_tensors = [make_tensor_value_info(name, onnx.TensorProto.FLOAT, shape)
                        for name, shape in zip(onnx_node.output, output_shapes)]

   print('output_tensors: ' + str([str(n) for n in output_tensors]))
   graph = make_graph([onnx_node], 'compute_graph', input_tensors, output_tensors)
   print('graph: ' + str(graph))
   model = make_model(graph, producer_name='ONNXNgraphImporter')
   model.opset_import[0].version = opset
   return model


x = np.array([[[[0., 1., 2., 3., 4.],  # (1, 1, 7, 5) input tensor
                [5., 6., 7., 8., 9.],
                [10., 11., 12., 13., 14.],
                [15., 16., 17., 18., 19.],
                [20., 21., 22., 23., 24.],
                [25., 26., 27., 28., 29.],
                [30., 31., 32., 33., 34.]]]]).astype(np.float32)

W = np.array([[[[1., 1., 1.],  # (1, 1, 3, 3) tensor for convolution weights
                [1., 1., 1.],
                [1., 1., 1.]]]]).astype(np.float32)

b = np.array([2]).astype(np.float32)



model = get_node_model('Conv', [(1,1,4,3)], x, W, b, kernel_shape=[3,3],
                        pads=[1,1,1,1], strides=[2,2])
onnx.save_model(model, 'conv_with_strides_padding_bias.onnx')
```

**Comment by postrational** [comment]
@arogowie-intel I don't know why you had a problem with your script. I ran the same code and got an ONNX model with proper inputs and outputs. Added fixed model.

**Comment by postrational** [comment]
I'm gettting compilation errors and warnings when trying to compile current version:
```
[  4%] Building CXX object src/ngraph/frontend/onnx_import/CMakeFiles/onnx_import.dir/op/conv.cpp.o
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.cpp:21:
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.hpp:19:
ngraph/src/ngraph/frontend/onnx_import/exceptions.hpp:32:44: error: invalid
      operands to binary expression ('const std::string' (aka 'const basic_string<char, char_traits<char>,
      allocator<char> >') and 'const char *')
                    : ngraph_error{op_name + " node (" + name + "): " + message}
                                   ~~~~~~~ ^ ~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:765:1: note: candidate template ignored: could not
      match 'reverse_iterator<type-parameter-0-0>' against 'char const[8]'
operator+(typename reverse_iterator<_Iter>::difference_type __n, const reverse_iterator<_Iter>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1184:1: note: candidate template ignored: could not
      match 'move_iterator<type-parameter-0-0>' against 'char const[8]'
operator+(typename move_iterator<_Iter>::difference_type __n, const move_iterator<_Iter>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1597:1: note: candidate template ignored: could not
      match '__wrap_iter<type-parameter-0-0>' against 'const char *'
operator+(typename __wrap_iter<_Iter>::difference_type __n,
^
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.cpp:21:
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.hpp:19:
ngraph/src/ngraph/frontend/onnx_import/exceptions.hpp:44:48: error: invalid
      operands to binary expression ('const std::string' (aka 'const basic_string<char, char_traits<char>,
      allocator<char> >') and 'const char *')
                        : ngraph_error{op_name + " node (" + name + "): " + message}
                                       ~~~~~~~ ^ ~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:765:1: note: candidate template ignored: could not
      match 'reverse_iterator<type-parameter-0-0>' against 'char const[8]'
operator+(typename reverse_iterator<_Iter>::difference_type __n, const reverse_iterator<_Iter>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1184:1: note: candidate template ignored: could not
      match 'move_iterator<type-parameter-0-0>' against 'char const[8]'
operator+(typename move_iterator<_Iter>::difference_type __n, const move_iterator<_Iter>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1597:1: note: candidate template ignored: could not
      match '__wrap_iter<type-parameter-0-0>' against 'const char *'
operator+(typename __wrap_iter<_Iter>::difference_type __n,
^
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.cpp:21:
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.hpp:20:
In file included from ngraph/src/ngraph/frontend/onnx_import/node.hpp:25:
In file included from ngraph/cmake-build-debug/src/ngraph/frontend/onnx_import/onnx.pb.h:9:
ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/stubs/common.h:215:29: warning:
      dynamic exception specifications are deprecated [-Wdeprecated-dynamic-exception-spec]
  virtual ~FatalException() throw();
                            ^~~~~~~
ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/stubs/common.h:215:29: note:
      use 'noexcept' instead
  virtual ~FatalException() throw();
                            ^~~~~~~
                            noexcept
ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/stubs/common.h:217:36: warning:
      dynamic exception specifications are deprecated [-Wdeprecated-dynamic-exception-spec]
  virtual const char* what() const throw();
                                   ^~~~~~~
ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/stubs/common.h:217:36: note:
      use 'noexcept' instead
  virtual const char* what() const throw();
                                   ^~~~~~~
                                   noexcept
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.cpp:21:
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.hpp:20:
In file included from ngraph/src/ngraph/frontend/onnx_import/node.hpp:25:
In file included from ngraph/cmake-build-debug/src/ngraph/frontend/onnx_import/onnx.pb.h:25:
In file included from ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/generated_message_table_driven.h:35:
ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/map_entry_lite.h:357:35: warning:
      declaration shadows a local variable [-Wshadow]
          typename Map::size_type size = map_->size();
                                  ^
ngraph/cmake-build-debug/protobuf/src/src/google/protobuf/map_entry_lite.h:352:13: note:
      previous declaration is here
        int size;
            ^
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.cpp:21:
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.hpp:21:
ngraph/src/ngraph/frontend/onnx_import/util/broadcasting.hpp:37:23: warning:
      parameter 'axis' not found in the function declaration [-Wdocumentation]
             * @param axis         The axis along which we want to replicate elements. The starting
                      ^~~~
ngraph/src/ngraph/frontend/onnx_import/util/broadcasting.hpp:37:23: note: did you
      mean 'start_match_axis'?
             * @param axis         The axis along which we want to replicate elements. The starting
                      ^~~~
                      start_match_axis
ngraph/src/ngraph/frontend/onnx_import/util/broadcasting.hpp:68:23: warning: no
      newline at end of file [-Wnewline-eof]
} // namespace  ngraph
                      ^
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.cpp:21:
In file included from ngraph/src/ngraph/frontend/onnx_import/op/conv.hpp:22:
ngraph/src/ngraph/frontend/onnx_import/util/conv_pool.hpp:53:37: warning: implicit
      conversion increases floating-point precision: 'float' to 'double' [-Wdouble-promotion]
                            return (static_cast<float>(dim) - 1.) / 2.;
                                    ^~~~~~~~~~~~~~~~~~~~~~~ ~
ngraph/src/ngraph/frontend/onnx_import/util/conv_pool.hpp:111:78: warning: use of
      old-style cast [-Wold-style-cast]
                    "strides", std::vector<std::size_t>(kernel_shape.size(), (std::size_t)1))};
                                                                             ^            ~
ngraph/src/ngraph/frontend/onnx_import/util/conv_pool.hpp:136:80: warning: use of
      old-style cast [-Wold-style-cast]
                    "dilations", std::vector<std::size_t>(kernel_shape.size(), (std::size_t)1))};
                                                                               ^            ~
ngraph/src/ngraph/frontend/onnx_import/util/conv_pool.hpp:171:74: warning: use of
      old-style cast [-Wold-style-cast]
                    pads = std::vector<std::size_t>(kernel_shape.size(), (std::size_t)0);
                                                                         ^            ~
```

**Comment by postrational** [comment]
This PR has grown too large. We’ll need to split it into smaller pieces.

**Comment by diyessi** [comment]
@postrational are you saying don't merge this? If so, please make it a changes requested.

**Comment by postrational** [comment]
This code was merged as a series of smaller PRs.

**Pull request closed without merging**